### ISSUE TITLE
bad `match_ref_pats` suggestion.

### ISSUE THREAD

**Issue by Lokathor** [rust-lang/rust-clippy#4834]
In [this line of this file of this commit](https://github.com/Lokathor/imagine/blob/d98b9bffeb05b985d456d992954bd61338fe1887/src/png.rs#L70), clippy says "you don't need to add `&` to all patterns", but because the `&` is actually changing the type of the binding, if you take _off_ that `&` then the type going into `png_u16` become wrong and the code doesn't compile.

**Comment by flip1995** [comment]
We shouldn't emit this lint if the match is on `&[T]`.

**Comment by ebroto** [comment]
The lint also suggests to dereference the scrutinee. Wouldn't that work as expected?
```rust
fn test() {
    let v: &[i32] = &[1, 2, 3, 4];

    #[allow(clippy::single_match)]
    match *v {
        [a0, a1, b0, b1] => {
            takes_arr([a0, b0]);
            takes_arr([a1, b1]);
        },
        _ => {},
    }
}

fn takes_arr(_: [i32; 2]) {}

fn main() {
    test();
}
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=747f2f198dba684320ec4b4199d25ddf))
